Add operatorChain()

simonbasle opened the pull request:
Second take at #1115, starts with the sequenceName commit + revert.
Third and fourth commits are the interesting ones.

@smaldini I don't think that skipping assembly ops in `parents()` or `actuals()` helps in any way.
Actually, while initially exploring your suggestion, delegating to their actual for `scan` would short-circuit the traversal.

Since the end goal is to expose a list/stream of the operators in the chain, I went completely in another direction:

 - add a `List<String> operatorChain()` method for this purpose. it assembles parents, this and actuals in a single list in assembly order
 - improve `operatorName()` of the assembly operators to display the stackline
 - add a `isMetaData()` marker method to `Scannable`, implemented by assembly operators
 - rely on this to collate operatorName between an assembly op and its op in `operatorChain`
 - mark all source Flux/Mono with `SourceProducer`, that is just implementing `Scannable` for the purpose of having an `operatorName` common to all sources

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+ *
+ * @author Simon Baslé
+ */
+class Traces {
```
Could be final

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+	 * @param stackTraceElements the array of {@link StackTraceElement} to convert to {@link String}
+	 * @return the string version of the stacktrace.
+	 */
+	public static String stackTraceToSanitizedString(StackTraceElement[] stackTraceElements) {
```
no need for public modifier

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+	 * @param stackTraceElements the array of {@link StackTraceElement} to convert to {@link String}
+	 * @return the string version of the stacktrace.
+	 */
+	public static String stackTraceToString(StackTraceElement[] stackTraceElements) {
```
no need for public modifier

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+	 * @param stackTraceRow the row to check
+	 * @return true if it should be sanitized out, false if it should be kept
+	 */
+	public static boolean shouldSanitize(String stackTraceRow) {
```
no need for public modifier

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+	 * @return a {@link String} representing operator and operator assembly site extracted
+	 * from the assembly stack trace.
+	 */
+	public static String extractOperatorAssemblyInformation(String source, boolean skipFirst) {
```
no need for public modifier

smaldini commented on reactor-core/src/main/java/reactor/core/publisher/Traces.java:
```diff
@@ -0,0 +1,223 @@
+	 * @return a {@link String} representing operator and operator assembly site extracted
+	 * from the assembly stack trace.
+	 */
+	public static String extractOperatorAssemblyInformation(String source) {
```
no need for public modifier

smaldini commented on reactor-core/src/main/java/reactor/core/Scannable.java:
```diff
@@ -380,28 +395,77 @@ default String name() {
+	 *
+	 * @return a {@link Stream} of {@link #operatorName()} for each discovered operator in the chain
+	 */
+	default Stream<String> operatorChain() {
```
I wonder if we should not keep a more neutral name (also for operatorName) in Scannable since it can apply to other dependency-like trace. 
WDYT about : dependencies() or stacktrace() ?

smaldini commented on reactor-core/src/main/java/reactor/core/Scannable.java:
```diff
@@ -380,28 +395,77 @@ default String name() {
-	 * @return the name of the first parent that has one defined (including this scannable)
+	 * @return the default operatorName of this operator
 	 */
 	default String operatorName() {
```
It's already in 3.1 but having now operatorChain made me realize that we could over time associate more and more Scannable with operators. I think this one is fine, the documentation might not be tho and should say: "Will format the name of this scannable accordingly if operator implementation, ex: ...."

simonbasle replied:
How about `stepName()` and `steps()`? `name()` would still represent an upstream search of a user-provided name for the sequence.

simonbasle commented:
@smaldini name changes implemented in `Scannable`:
 - kept `name()` for the sequence-wide user-provided naming (actually, lookup in upstreams)
 - changed `operatorChain()` to `steps()`, as a `Stream<String>`
 - kept `operatorName()`, **deprecated**
 - added `stepName()` as an alias (and ultimately replacement) of `operatorName()`. it is aliasing it so that implementations that override `operatorName()` still have their override invoked whenever `stepName()` is called.

Alternatively we could remove `operatorName()` entirely, renaming it to `stepName()`. Note that for japicmp it would be the first API breaking change of 3.2.0 so far.

smaldini commented:
Nice @simonbasle good to rebase/merge and iterate on the deprec. question later.
